### ISSUE TITLE
Much more flexible token authentication

### ISSUE THREAD

**Issue by djmitche** [mozilla/build-relengapi#163]
(Bugzilla: https://bugzilla.mozilla.org/show_bug.cgi?id=1132957)
# Temporary Tokens

The use-case is a taskcluster scheduler which needs to enable very limited access for a single task, in the form of a token that has a higher likelihood of being revealed (via verbose logging or because it's passed to untrusted try-level code).

These tokens should
 * be issuable primarily via API
 * have a limited lifetime
 * be revokable (when the job ends)
 * specify arbitrary, trusted additional metadata that's available to the API (e.g., "can upload performance statistics for test suite XXX for revision YYY")

# User Tokens

The existing tokens are permanent and not tied to a particular person -- so they don't go away when that person does.

We need another form of token, which
* anyone who can authenticate to RelengAPI can issue,
* stops working when the issuer's account stops working, and
* appears internally as a login by that user (e.g., "Human:dmitchell@mozilla.com")
* otherwise does not expire

# OAuth

The idea here is to allow other apps at Mozilla to programmatically request a subset of a user's permissions on RelengAPI.  Similar to the way lots of apps in the real world connect to users' GitHub accounts.  In fact, the process should be modelled on GitHub as closely as possible, to eliminate user surprise.

On GitHub, application registration looks like this:

![tokens-dev-apps](https://cloud.githubusercontent.com/assets/28673/5455018/bb949356-8505-11e4-8ba4-bbe205a85b6b.png)
![tokens-register](https://cloud.githubusercontent.com/assets/28673/5455017/bb909314-8505-11e4-8b5e-d061811c9c04.png)

The authorization ask from users looks like this:

![tokens-authorize](https://cloud.githubusercontent.com/assets/28673/5455056/222a2c52-8506-11e4-90f6-4cf6026a060b.png)

Once apps are authorized, they are displayed like this:

![tokens-authorized-apps](https://cloud.githubusercontent.com/assets/28673/5455034/ee482ab0-8505-11e4-91e5-0d9422ea8e73.png)

GitHub's docs: https://developer.github.com/v3/oauth/

**Comment by djmitche** [comment]
## Requests

Requests can be authenticated by:

Input | User Type | Permissions | Duration | Notes
------|-----------|-------------|----------|-------
browser cookie | HumanUser | from authz lookup of user | refreshed on every request |
permanent token (`prm`) | TokenUser | perms granted to token | unlimited | revokable
temporary token (`tmp`) | TokenUser | perms granted to token | limited | non-revokable
user token (`usr`) | HumanUser | intersection of user's authz perms and token perms | unlimited | TokenUser is OK too; user must be valid
app access token (`acc`) | AppUser | token perms | limited | non-revokable
app refresh token (`ref`) | AppUser | request app access token | unlimited | revokable; user must be valid and have required perms
app client token (`cli`) | ClientUser | request app/refresh tokens | unlimited | requires a grant, too; revokable

All but the first of these come in the form of a bearer token in JWS format.  The [JWT](http://jwt.io) claims are:

 * `iss` - `ra2`, for RelengAPI JWT version 2
 * `typ` - token type (see above)
 * limited-duration, non-revokable tokens only
   * `exp` and `nbf` - used for limited-duration / non-revokable tokens
   * `prm` - list of permissions the token represents
 * unlimited-duration, revokable tokens only
   * `jti` - `t` + token id (from the tokens table), used for unlimited-duration / revokable tokens
   * user-linked tokens (`usr`,  `ref`):
     * `sub` - user email
   * app-linked tokens (`cli`)
     * `sub` - `c` + client id (from clients table)

Note that the limited-duration tokens contain sufficient information to validate a request without consulting the database.  Unlimited-duration tokens require a database lookup.

XXX: include a salt?

## Tables

 * `tokens` (only `prm`, `usr`, `ref`, and `cli` tokens -- those with a `jti`)
   * `id`
   * `typ` - redundant to the jwt, but helpful for knowing what fields to expect
   * `permissions` - comma-separated
   * `description` (`prm`, `usr`, `cli` only)
   * `user` (`usr`, `ref` only) - used to revoke (by deleting the row) when user is disabled
 * `clients`
   * `id`
   * `name`
   * `is_confidential` (True -> server-side profile; False -> client-side or native profile)
   * `description`
   * `homepage`
   * `redirect_uri`
   * `creator` - user email, or NULL for a permanent client
   * note: no secret -- secret is a `cli` token (and can thus, in principle, be re-issued)
 * `grants` - stored in cache only

## REST API

### Endpoints

All of these are under `/tokenauth`.

* `/tokens` - list all visible tokens; with `?typ=`, filter by type
* `POST /tokens` - issue a token of any allowed type
* `/tokens/<token_id>` - get any revokable token
* `DELETE /tokens/<token_id>` - revoke a revokable token
* `POST /tokens/query` - interpret a token (including data from the DB if appropriate)
* `/clients` - list all clients
* `POST /clients` - create a new client
* `/clients/<client_id>` - get a client
* `DELETE /clients/<client_id>` - delete a client

### Types

 * Token (all fields except `typ` are optional)
   * `id`
   * `typ`
   * `not_before`
   * `expires`
   * `permissions`
   * `description`
   * `user`
   * `client_id`
   * `token` - token text; only included on issue

 * Client
   * `id`
   * `name`
   * `is_confidential` (True -> server-side profile; False -> client-side or native profile)
   * `description`
   * `homepage`
   * `redirect_uri`
   * `creator` - user email, or NULL for a permanent client
   * `secret` - token text; only included on issue (this is actually a token)

# Self-revocation

All users have the 'base.tokens.revoke-self` permission, and can grant that to a token.  The revocation endpoint will allow an unlimited-duration token with this permission to revoke itself.

# Clients and Client Secrets

A newly-minted client is returned with a "client secret".  This is actually a token with type `cli`, and in principle could be regenerated without changing the client id by posting an appropriate Token to `/tokens`.  Permissions on this are tricky, so it doesn't need to be implemented immediately.

**Comment by djmitche** [comment]
# TODO

 * [ ] examine existing implementations and see what "easy" ways we can make the flow more compatible
   * GitHub
   * Flask-oAuth
   * Some browser-side oAuth client lib?
 * [ ] determine permissions model for the new endpoints
   * "my own" vs. "any" for most token types
   * global client admin privs, with permission to "promote" a client from user-owned to permanent
 * [ ] implement temporary tokens
 * [ ] implement API for per-user tokens
 * [ ] implement UI for per-user tokens
 * [ ] implement monitoring of user permissions and disabling associated tokens
 * [ ] implement OAuth
 * [ ] implement disabling clients when owning users are gone

**Comment by Callek** [comment]
fwiw, I don't see a strong value in making temp tokens non-revokable. I can imagine reasons for every type of token to be revokable. Even if some types are unlikely to be revoked commonly.

**Comment by djmitche** [comment]
The issue isn't that there's value in a token being non-revokable.  It's that making a token revokable requires storing state about that token, and thus requires a database hit on every use of that token.  Temporary tokens are likely to see a lot of usage, so avoiding this hit will be a big improvement in scalability.  They also have limited permissions, and will last hours, not weeks, so revocation shouldn't be necessary.

**Comment by Callek** [comment]
Just to say it, if there is a no-UI way to revoke a temp token, I'd really really want some sort of documentation on how we can achieve a revoke if there is ever a sec issue with one existing somewhere. I don't want to be in a position where we need to shutdown all of relengapi just due to a single token being a sec issue.  (however unlikely, -- I like to plan that the unlikely can happen when thinking about security)

**Comment by djmitche** [comment]
# Existing APIs

## GitHub OAUth2 API

* redirect to `https://github.com/login/oauth/authorize` with
  * `client_id`
  * `redirect_uri`
  * `scope`
  * `state` (CSRF prevention)
* Browser bounced back to `redirect_uri` with
  * `code` (grant)
  * `state` (should match)
* server-side POST to `https://github.com/login/oauth/access_token` with
  * `client_id`
  * `client_secret`
  * `code` (from redirect)
  * `redirect_uri`
* response has
  * `access_token`
  * `token_type` (bearer)
  * `scope`
* make requests with
  * `Authorization: token $access_token`

So, no refresh tokens are involved.

**Comment by djmitche** [comment]
Per RFC6749, it looks like the refresh token takes the place of a grant code in fetching a new access token when the previous had expired.

**Comment by djmitche** [comment]
In implementing OAuth, I'd like to make a nice checklist of stuff to verify -- preferably automatically -- about the implementation.  I think that would come from the MUST and SHOULD's in RFC 6749 and a selection of the more concrete countermeasures in RFC6819.  A few random bits I need to make sure are present:

 * register a single redirect_uri per client and validate it on every request; require a complete match (not just prefix)
 * never use a redirect_uri if it can't be validated against a registered client
 * require that the redirect_uri be protected by https (or at least warn loudly if it is not) [RFC 6819 4.4.1.6](http://tools.ietf.org/html/rfc6819#section-4.4.1.6)

**Comment by djmitche** [comment]
Based on http://pure.ltu.se/portal/en/studentthesis/securing-public-apis-using-oauth-and-oauthlib%2893e405ff-edde-4b60-bc80-8d237553807a%29.html there are a few additional things to worry about:
 * using a DB row id in the user-visible portion of the JWT is both too little entropy, and too much information to disclose (it facilitates enumeration, although such enumeration would also require the signing secret).  Instead we should use some kind of randomly-generated unique identifier.

**Comment by djmitche** [comment]
Flask-OAuthlib has some serious issues, and we won't be using it.

I implemented a toy server, client, and test suite in https://github.com/djmitche/oauth-toy

The test suite is a prototype for a way to easily test that the implementation is meeting all of the MUSTs and whatnot.  The idea is to realistically walk through the entire protocol, then adjust the behavior of the client or the user and make sure the server responds appropriately.

**Comment by djmitche** [comment]
§2.3.1 specifies that servers should support "Basic" auth for clients: base64("$client_id:$client_secret").  That's not compatible with the existing token auth stuff, so we won't be authenticating clients with tokens, but just using stored hashed passwords.

**Comment by djmitche** [comment]
I talked to @jvehent and @gdestuynder today about the available use-cases for OAuth support, and the complexity of implementing it.  The email was

---

Hey, so I'm having second thoughts.

First, the really useful model for OAuth2 is when a user is putting
their *own* data at risk by authorizing an application.  For example,
if I authorize a malicious app to access my friends list in Facebook,
the worst the app can do is grab the info about my friends and spam
them or whatever -- fb's infrastructure is not in peril.  RelengAPI
has very little that is "user-owned" -- basically just try builds.
Buildslaves, version-control histories, clobbers, etc. are all shared
resources, and making an easy way to authorize third-party code to
manipulate those resources is risky and has low reward.

Second, I'm not particularly happy with any of the libraries out
there.  OAuth2 is a pretty simple protocol, but of course simple to
understand and simple to implement are different things.  I had hoped
to find a small, elegant library with a solid, reliable
implementation.  No dice.  Oauthlib seems well-implemented, but it's
not documented, which makes me worry that I'm mis-using it in some
subtle way.

Third, the immediate demands in the auth/authz realm are:
 * short-term authorization of a taskcluster job to access tooltool
 * replacement of taskcluster's use of persona for auth
 * control of releng processes via other Mozilla tools like
treeherder, treestatus

For none of those is OAuth2 a good choice (and it'd be a particularly
BAD choice for replacing persona)

---

We agreed to set aside work on OAuth for the moment as high-risk and low-reward.  The question of how to make relengapi a "remixable" tool remains open -- we batted around everything from signing API requests using OpenPGP.js to sharing SAML assertions.  But at the moment there's not a lot in the API that needs protecting.  We divided up the API along two axes as follows:

 * public data -- things like build status, version control, etc.  No need for access control here.
 * private data -- there will likely be very little of this in relengapi

and

 * user-specific data and methods -- *my* try builds, *my* loaner, etc.
 * non-user-specific data and methods -- buildslave information, non-try builds, clobberer, etc.

OAuth is really only a good option for private data, and for user-specific data and methods.  And it's not clear we'll have much of either one.

There's more work to do for tokens, specifically around expiring user tokens when users leave, so I'll take care of that now.
